pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -270,6 +270,9 @@ private void processSingleValue(int numDocs, int[] dictIds, int[] outGroupIds) {
         if (!_flags[outGroupIds[i]]) {
           _numKeys++;
           _flags[outGroupIds[i]] = true;
+          if (_numKeys == _globalGroupIdUpperBound) {
```

Comment:
Seems like the change was intended for conditional exit for low cardinality. I don't think that's what is happening here ?

Reply:
This loop will only exit early when the cardinality is low.

Reply:
We should also add one check before the for loop.

Out of the scope of this PR, in `processSingleValueGeneric()` and `processMultiValue()`, should we move `_flags[groupId] = true` into the if block?

Reply:
@Jackie-Jiang I'm not sure I follow - could you be more explicit please?

Reply:
I think I got it, please see the latest commit